tdwii_plus_examples/upswatchnactionscu.py

```
@@ -0,0 +1,273 @@
+            action_info.DeletionLock = str(deletion_lock).upper()
+            if matching_keys is not None:
+                instance_uid = UPSFilteredGlobalSubscriptionInstance
+                for elem in matching_keys:
+                    action_info.add(elem)
+        rsp_status = Dataset()
+        self.logger.debug("Sending N-ACTION request")
+        rsp_status, action_reply = assoc.send_n_action(
+            dataset=action_info,
+            action_type=action_type_id,
+            class_uid=UnifiedProcedureStepPush,
+            instance_uid=instance_uid,
+        )
+
+        try:
+            self.status = self._handle_response(rsp_status, action_reply)
+        except ResponseWarning as warning:
+            self.logger.warning(f"{warning.message} (Status Code: {warning.status_code})")
+            raise
+        except ResponseError as error:
+            self.logger.error(f"{error.message} (Status Code: {error.status_code})")
+            raise
+        except ResponseUnknown as unknown:
+            self.logger.error(f"{unknown.message} (Status Code: {unknown.status_code})")
+            raise
+        finally:
+            self.assoc.release()
+            self.logger.debug("Association released")
+
+    def _toggle_subscription(self, action_type_id, instance_uid=UPSGlobalSubscriptionInstance, lock=False, matching_keys=None):
+        safe_to_proceed = False
+        try:
+            self._associate()
+            safe_to_proceed = True
+        except AssociationError as error:
+            self.logger.error(str(error))
+            raise
+        except ContextWarning as warning:
+            accepted_sop_names = [f"[{UID(uid).name}]" for uid in warning.accepted_sop_classes]
+            self.logger.warning(f"{warning} - Accepted Transfer Syntaxes: {', '.join(accepted_sop_names)}")
+            if "[Unified Procedure Step - Watch SOP Class]" in accepted_sop_names:
+                safe_to_proceed = True
+            raise
```

if it's safe to proceed, I don't think we should raise an exception.
The finally will "do the right thing", but unless a function caller is prepared to handle an exception that is more of a warning, I think the application will act as if there was a serious failure.  And if the programmer gets used to catching exceptions and not really acting on them (because they get warnings in exceptions), then they won't respond with a fail fast/fail safe when they should.

---

Custom exceptions replaced by return of named tuple.